### PR TITLE
ADR for improving FE developer experience

### PR DESCRIPTION
**Pull request by fhenrich33**

### CODE REVIEW

**Comment by fhenrich33** [comment]
[2.0 FE: fix automated linting, formatting and type checking in FE before commiting](https://trello.com/c/gtV5lYon/1428-20-fe-fix-automated-linting-formatting-and-type-checking-in-fe-before-commiting)

File: docs/adr/adr_improved_frontend_dx.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+Decision Deadline: As soon as possible
+
+Author: [Felipe](https://github.com/fhenrich33)
+
+## Status
+
+Proposed, Implementing
+
+## Context
+Frontend development experience in its current form is quite lacking. Imposes a lot of artificial difficulties, tooling doesn't work as expected, and adds friction to our FE contributors.
```
Comment (aerinsol) [inline docs/adr/adr_improved_frontend_dx.md:11]: "quite lacking" and "artificial difficulties" needs to be more specific

File: docs/adr/adr_improved_frontend_dx.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+We want to make FE development better.
+
+## Decision drivers
+
+1. We must make FE onboarding and development seamless.
```
Comment (aerinsol) [inline docs/adr/adr_improved_frontend_dx.md:17]: Seamless, as in independent of back end tooling?

File: docs/adr/adr_improved_frontend_dx.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+2. Implementation must play well with the current FE setup (React/Vite).
+3. Implementation must work well with our CI/CD pipeline.
+4. We must NOT interfere with BE tooling and development.
+
+## Considered options
+
+The only difference between proposed options is how to manage the monorepo, otherwise, there's a lot of overlap on proposed changes:
+- **[pnpm](https://pnpm.io/):** The "new" unwritten standard for package management--faster than both `npm` and `yarn`, and more forgiving on how it manages and caches `node_modules`. Will replace `yarn`.
+- **[husky](https://typicode.github.io/husky/) and [lint-staged](https://github.com/lint-staged/lint-staged):** Improved git hook configuration, also an unwritten standard in the FE ecosystem. Will replace our manual git hook setup.
```
Comment (aerinsol) [inline docs/adr/adr_improved_frontend_dx.md:26]: Does "unwritten standard" here mean "the de facto standard"?

File: docs/adr/adr_improved_frontend_dx.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+## Context
+Frontend development experience in its current form is quite lacking. Imposes a lot of artificial difficulties, tooling doesn't work as expected, and adds friction to our FE contributors.
+
+We want to make FE development better.
+
+## Decision drivers
+
+1. We must make FE onboarding and development seamless.
+2. Implementation must play well with the current FE setup (React/Vite).
```
Comment (aerinsol) [inline docs/adr/adr_improved_frontend_dx.md:18]: There are two implementation considerations we always consider with things like this, which is maintainability and longevity, since our team is rather small and can't afford to regularly upgrade. Ideally it would be 5 year longevity with minimum maintainability issues. Would be good if this could be addressed here and with all other tooling related ADRs - you'll see I made a similar comment on @pylipp prior ADRs.

File: docs/adr/adr_improved_frontend_dx.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+The only difference between proposed options is how to manage the monorepo, otherwise, there's a lot of overlap on proposed changes:
+- **[pnpm](https://pnpm.io/):** The "new" unwritten standard for package management--faster than both `npm` and `yarn`, and more forgiving on how it manages and caches `node_modules`. Will replace `yarn`.
+- **[husky](https://typicode.github.io/husky/) and [lint-staged](https://github.com/lint-staged/lint-staged):** Improved git hook configuration, also an unwritten standard in the FE ecosystem. Will replace our manual git hook setup.
+
+Current tooling like `Vite`, `TypeScript`, `ESLint`, and `Prettier` remains the same, and are expected to work flawlessly with these.
+
+Monorepo management:
+
+1. **[Turborepo](https://turbo.build/repo/docs):** Does a lot of lifting for us automatically, is content aware so it only works on relevant files, caches results of deterministic tasks locally ([and remotely](https://turbo.build/repo/docs/core-concepts/remote-caching), if we have a need for it in the future), and works seamlessly with current and suggested tooling.
```
Comment (aerinsol) [inline docs/adr/adr_improved_frontend_dx.md:32]: Lifting like what?

File: docs/adr/adr_improved_frontend_dx.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+2. **[pnpm workspaces](https://pnpm.io/workspaces):** A safer option, at the cost of doing what `Turborepo` does in a more laborious and manual way.
+
+## Decision
+The considered options should be an improvement over the current FE setup, both in my own experience and what can be observed in our industry at the time of this writing. Most of these are expected to be installed locally or already heavily used by our FE developers. These also shouldn't interfere with the BE developer experience in any way.
+
+Another highly recommended action: change the required git hook from `pre-commit` to `pre-push`. Let plugins, tooling and IDE/text editor of choice take care of chores like linting and formatting locally in order to remove friction, and rely on CI/CD for the final say on correctness and confidence.
```
Comment (aerinsol) [inline docs/adr/adr_improved_frontend_dx.md:38]: Prepush, why?

**Review by aerinsol — CHANGES_REQUESTED** [review]

File: docs/adr/adr_improved_frontend_dx.md (hunk 3 above)
Comment (fhenrich33) [reply docs/adr/adr_improved_frontend_dx.md:26]: I would not say a de facto standard since we don't even have a standard library for Javascript. But since it's very commonplace, that's why I picked these words specifically.

File: docs/adr/adr_improved_frontend_dx.md (hunk 6 above)
Comment (fhenrich33) [reply docs/adr/adr_improved_frontend_dx.md:38]: At least in my experience, leveraging your IDE, LSP (language server protocol) and tooling to automate tasks that otherwise would be attached to the pre-commit hook leads to a much better development experience, in comparison with doing these on pre-commits. The reason is that as I commit my code, the pre-commit is enforcing something that's either not relevant at the moment or might break my train of thought, while the alternative only watches and does tasks on my files contextually.

Of course, let me know if you have concerns in that regard.

File: docs/adr/adr_improved_frontend_dx.md (hunk 5 above)
Comment (fhenrich33) [reply docs/adr/adr_improved_frontend_dx.md:32]: Should have been more clear here, but everything: type checking, formatting, testing, building, etc.

File: docs/adr/adr_improved_frontend_dx.md (hunk 2 above)
Comment (fhenrich33) [reply docs/adr/adr_improved_frontend_dx.md:17]: I have a bullet point for that, but yes. And also shouldn't be as wonky as it is now e.g. having a shell script to invoke Yarn, which for some reason doesn't work with my local installation.

File: docs/adr/adr_improved_frontend_dx.md (hunk 1 above)
Comment (fhenrich33) [reply docs/adr/adr_improved_frontend_dx.md:11]: I have mentioned it briefly here: https://github.com/boxwise/boxtribute/pull/1400/files#r1663184123

But although I can get the app up and running quickly, my local environment doesn't identify things and play well with the code as I expect it to work. TS, ESlint, and Prettier don't work at all without a lot of fiddling. 

I'll accentuate that in the ADR.

File: docs/adr/adr_improved_frontend_dx.md (hunk 4 above)
Comment (fhenrich33) [reply docs/adr/adr_improved_frontend_dx.md:20]: Are you looking for a more specific way of wording and addressing these concerns?

I ask because while I've written this ADR and the accompanying implementation (TBR soon) with maintainability and longevity in mind, I might not have eased concerns regarding those very valid points.

Comment (aerinsol) [reply docs/adr/adr_improved_frontend_dx.md:20]: Yes, sorry let me be more clear about the expectations here. Imperfect draft ADRs are certainly better than none at all, but the goal here is to fully capture the context, including the dimensions that were and were not considered. For environment changes in particular, I find it helpful to explicitly call out maintainability, scalability and power and how the options considered stack up. This is an old but still valid decision around engineering excellence to not necessarily go with the flashiest new thing, but the one that is the most maintainable yet powerful. 


An example of how this was recently handled for the FE bundler: https://github.com/boxwise/boxtribute/blob/master/docs%2Fadr%2Fadr_fe_bundler.md
This is a relatively "rough" writeup, but note that it explicitly covers maintainability and reliability. 

An more polished example from the BE where we explicitly cover reliability and power, and ultimately chose not to go with the gold standard: https://github.com/boxwise/boxtribute/blob/master/docs%2Fadr%2Fadr_Python-ORM.md I will note that despite its more polished state, it would be improved if maintainability waa covered in the same was as in the Vite writeup.

Comment (pylipp) [reply docs/adr/adr_improved_frontend_dx.md:20]: > Implementation must play well with the current FE setup (React/Vite).

I'd like to add that it's probably preferred to avoid major friction when switching the tool stack (can't assess it myself from the proposed selection of tools), and it might be beneficial to do incremental updates if possible (again, can't assess that).

File: docs/adr/adr_improved_frontend_dx.md [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+Author: [Felipe](https://github.com/fhenrich33)
+
+## Status
+
+Proposed, Implementing
+
+## Context
+Frontend development experience in its current form is quite lacking. It imposes a lot of artificial difficulties, tooling doesn't work as expected and adds friction to our FE contributors.
+
+Local installation of packages and tooling might not play well with the codebase. Things like TS, ESlint, and Prettier might not work at all without a lot of fiddling.
```
Comment (pylipp) [inline docs/adr/adr_improved_frontend_dx.md:13]: Do @HaGuesto @vahidbazzaz @MaikNeubert currently work with the "old" FE tooling? Just curious if they experienced similar fiddling issues :sweat_smile:

File: docs/adr/adr_improved_frontend_dx.md [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+Local installation of packages and tooling might not play well with the codebase. Things like TS, ESlint, and Prettier might not work at all without a lot of fiddling. 
+
+We want to make FE development better.
+
+## Decision drivers
+
+1. We must make FE onboarding and development seamless.
+2. Implementation must play well with the current FE setup (React/Vite).
+3. To add to the prior point: it must be maintainable with minimal hassle in the long term.
+4. Implementation must work well with our CI/CD pipeline.
```
Comment (pylipp) [inline docs/adr/adr_improved_frontend_dx.md:22]: if not considered / included here yet: must work with dependabot

File: docs/adr/adr_improved_frontend_dx.md (hunk 3 above)
Comment (pylipp) [reply docs/adr/adr_improved_frontend_dx.md:29]: What's the manual git hook setup?

File: docs/adr/adr_improved_frontend_dx.md [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+5. We must NOT interfere with BE tooling and development.
```
Comment (pylipp) [inline docs/adr/adr_improved_frontend_dx.md:23]: There's one overlap of BE and FE in terms of tools: the GraphQL schema is currently being defined in `.graphql` files [here](https://github.com/boxwise/boxtribute/tree/master/back/boxtribute_server/graph_ql/definitions). We use the `graphql-codegen` tool to create TS type definitions from these for the FE.

As a BE dev I update/extend the schema files and want the TS type definitions to be aligned. If FE tooling setup is simple I would install it (if not, the current situation becomes reversed, slightly). Alternatively we could implement a bot that operates on PRs and applies potential TS type changes back to the current PR branch. What do you think is a good middle-ground solution?

**Review by pylipp — COMMENTED** [review]
Thanks for the thorough write-up @fhenrich33 I'm sorry that the current setup brought you hassle. I agree that using a single, Python-centered tool (`pre-commit`) poses hindrance for a simple FE DX due to its requirements (Python installation, venv, ...). I'm grateful for you to bring in your FE dev experience to find a working solution!

File: docs/adr/adr_improved_frontend_dx.md (hunk 3 above)
Comment (fhenrich33) [reply docs/adr/adr_improved_frontend_dx.md:29]: Sorry, poor choice of words on my end: I meant about the hand-rolled scripts like the one to invoke Yarn.

File: docs/adr/adr_improved_frontend_dx.md (hunk 4 above)
Comment (fhenrich33) [reply docs/adr/adr_improved_frontend_dx.md:20]: @aerinsol I see. I'll make sure to expand on maintainability and scalability.

@pylipp Related to that, I picked these pieces as they are rather unobtrusive and play nice with what we currently have.

File: docs/adr/adr_improved_frontend_dx.md (hunk 3 above)
Comment (pylipp) [reply docs/adr/adr_improved_frontend_dx.md:29]: That's not ideal indeed 😅

File: docs/adr/adr_improved_frontend_dx.md (hunk 7 above)
Comment (aerinsol) [reply docs/adr/adr_improved_frontend_dx.md:13]: also tagging @jlin95 since she recently installed our codebase :)

File: docs/adr/adr_improved_frontend_dx.md (hunk 4 above)
Comment (aerinsol) [reply docs/adr/adr_improved_frontend_dx.md:20]: @fhenrich33 great, if you've assessed it from a compatibility and upgradability standpoint, would be nice to see that explicitly called out. I would probably classify upgradability as a subset of maintainability.

File: docs/adr/adr_improved_frontend_dx.md (hunk 1 above)
Comment (aerinsol) [reply docs/adr/adr_improved_frontend_dx.md]: @jlin95

File: docs/adr/adr_improved_frontend_dx.md (hunk 7 above)
Comment (jlin95) [reply docs/adr/adr_improved_frontend_dx.md:13]: Some sorts of issues I saw:
`Cannot find module '@chakra-ui/react' or its corresponding type declarations.ts(2307)` in ui component files.
We do have the prettier and eslint config files in the codebase but it's not getting picked up properly by the IDE when you make changes that violate rules specified in those files.

File: docs/adr/adr_improved_frontend_dx.md (hunk 2 above)
Comment (jlin95) [reply docs/adr/adr_improved_frontend_dx.md:19]: @fhenrich33 out of curiosity why do you need a shell script to invoke Yarn (and by extension) what kind of local installation set up do you have? 
Running `docker compose up` after managing the env vars did the trick (at least on my machine) so I am wondering what sort of errors you're getting.

File: docs/adr/adr_improved_frontend_dx.md (hunk 9 above)
Comment (jlin95) [reply docs/adr/adr_improved_frontend_dx.md:23]: I think that is a reasonable solution to try out in the MVP build @fhenrich33 what do you think?

File: docs/adr/adr_improved_frontend_dx.md (hunk 1 above)
Comment (jlin95) [reply docs/adr/adr_improved_frontend_dx.md]: I replied in one of the discussion threads

**Review by jlin95 — COMMENTED** [review]
Left some comments specifically around the install setup

**Comment by aerinsol** [comment]
> Left some comments specifically around the install setup

Do you mostly agree or disagree with the recommendations? Can you summarize / share your overall thoughts.

**Comment by jlin95** [comment]
> > Left some comments specifically around the install setup
> 
> Do you mostly agree or disagree with the recommendations? Anything you want to add? Would like to see you summarize / share your overall thoughts.



> > Left some comments specifically around the install setup
> 
> Do you mostly agree or disagree with the recommendations? Anything you want to add? Would like to see you summarize / share your overall thoughts.

Overall I see this ADR in conjunction with the upcoming one that Felipe will work on regarding the MVP. 
I don't see anything I would disagree with, except for the rather strongly worded decision driver that 
`These also shouldn't interfere with the BE developer experience in any way.` 
-> is there a particular use case one might foresee our FE set up conflicting or impeding BE development?

**Comment by HaGuesto** [comment]
[Dropapp: add menue forwarding to Statviz Dashboard](https://trello.com/c/qycJolkU/1427-dropapp-add-menue-forwarding-to-statviz-dashboard)

File: docs/adr/adr_improved_frontend_dx.md (hunk 2 above)
Comment (fhenrich33) [reply docs/adr/adr_improved_frontend_dx.md:19]: @jlin95 this script is used on our pre-commit setup. But since these live inside Docker, we don't have any of the expected linting, typing, and such from our local setup. As you have experienced here: https://github.com/boxwise/boxtribute/pull/1400#discussion_r1667623496

File: docs/adr/adr_improved_frontend_dx.md [hunk 10]
```diff
@@ -1,5 +1,5 @@
 # ADR: Improved Frontend Developer Experience
-Decision Deadline: As soon as possible
+Decision Deadline: Next week (2024-07-15)
 
 Author: [Felipe](https://github.com/fhenrich33)
```
Comment (aerinsol) [inline docs/adr/adr_improved_frontend_dx.md:5]: Add: "Reviewed by: Roanna, Jia Lin, Philipp"

**Review by aerinsol — APPROVED** [review]
Happy with the proposed changes, let's see how the FE MVP works and then revisit with some comments from the team.

**Comment by aerinsol** [comment]
[EPIC: Frontend Developer Experience Improvements](https://trello.com/c/KN8cuKTK/1438-epic-frontend-developer-experience-improvements)
